### PR TITLE
[WIP] nwsapi

### PR DESCRIPTION
**Pull request by domenic**
@dperini: Here is my latest attempting at integrating nwsapi. I get a number of test failures:

- https://github.com/jsdom/jsdom/blob/4884fb9903a8ca9a371f643609de084e81592d2e/test/web-platform-tests/to-upstream/dom/nodes/ParentNode-querySelector-escapes.html#L95 fails to match
- https://github.com/jsdom/jsdom/blob/4884fb9903a8ca9a371f643609de084e81592d2e/test/web-platform-tests/to-upstream/dom/nodes/ParentNode-querySelector-escapes.html#L118 causes an exception
- https://github.com/jsdom/jsdom/blob/4884fb9903a8ca9a371f643609de084e81592d2e/test/web-platform-tests/to-upstream/dom/nodes/ParentNode-querySelector-escapes.html#L119 causes an exception
- https://github.com/jsdom/jsdom/blob/4884fb9903a8ca9a371f643609de084e81592d2e/test/web-platform-tests/to-upstream/dom/nodes/svg-template-querySelector.html#L20 returns null instead of matching
- https://github.com/jsdom/jsdom/blob/4884fb9903a8ca9a371f643609de084e81592d2e/test/web-platform-tests/to-upstream/svg/element-svg.html#L48 returns null instead of matching

I think there are a number more but that's what I started with. Let me know if there's anything else I can do to help debug. You should be able to debug by cloning this branch, running `yarn install`, then `yarn test`.

### CODE REVIEW

**Comment by dperini** [comment]
@domenic
thank you for the help and the initial attempt to switch to "nwsapi".
I will have a look at those errors asap.

File: lib/jsdom/living/helpers/selectors.js [hunk 1]
```diff
@@ -1,17 +1,25 @@
 "use strict";
 
 const idlUtils = require("../generated/utils");
-const nwmatcher = require("nwmatcher/src/nwmatcher-noqsa");
+const nwsapi = require("nwsapi"); // TODO: rename nwmatcher -> nwsapi everywhere
+const DOMException = require("domexception");
+
+// TODO: consider making this work, or getting nwsapi to switch to a different way of detecting XML documents
+function dummyXMLDocument() { }
```
[the remark this answers is not in the document]
Comment (dperini) [reply lib/jsdom/living/helpers/selectors.js]: @Zirro 
that's correct, I notice the additional check and workaround (and the comment for me) so I changed the detection into a simpler one to make it an easier integration and avoid a DOMException. Hope the check is strong enough, any suggestion is welcome.

**Comment by dperini** [comment]
@domenic 
I was able to fix the css-escaping issue in my "nwsapi" repository but I need to push a few extra commits. For the SVG errors I believe there is a problem in the tests ... I checked in the browsers and I see errors in one of the two tests you linked above.

File: lib/jsdom/living/helpers/selectors.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (domenic) [reply lib/jsdom/living/helpers/selectors.js]: Ohh, that's great, checking content-type seems perfect.

**Comment by dperini** [comment]
I just completed the upload of `nwsapi` 2.0.0beta1 on npm with the latest changes and fixes.
Check it out and see if all the reported errors have been fixed and jsdom test are passing.
I would say i am almost done with fixing errors and adding  features (those I know).
I will now start with a few optimization steps.

**Comment by Zirro** [comment]
I have tested this branch with the new version of nwsapi and it does indeed resolve the escaping-issues. I've also spent some time debugging the two SVG tests, and I believe that they are similar in cause to https://github.com/jsdom/jsdom/issues/1986.

In `svg-template-querySelector.html`, it seems like the path for elements within a DocumentFragment expects an uppercase `nodeName` as HTML elements have, even if they are outside the HTML namespace. It's still matching `<svg>` as intended in a regular Document, though.

I've not looked too closely at `element-svg.html` yet, but I suspect a similar cause. In this case the parent is a regular Document, but the selector (`svg#second`) matches by the element name + an ID. If I change it to just `#second` - or make the element a `<div>`  and change the selector to `div#second` - the test passes. There is presumably an issue when the name of an element outside the HTML namespace is combined with the remaining selector.

There is also a failure in `url/data-uri-fragment.html`. I see that it uses `:matches(:target)`, but I've not done testing to determine whether the actual problem is in nwsapi or jsdom yet.

Finally I just want to say thanks for working with us to resolve all these cases, @dperini. The work that's gone into the design of nwsapi - and the performance it can reach - is nothing short of remarkable.

**Comment by dperini** [comment]
@Zirro thank you for the support and the help, I am looking into to the mixed-namespace issues you noticed (mixed upper/lower cases of tag names).
Unfortunately none of `lookupNamespaceURI()`, `lookupPrefix()` and `isDefaultNamespace()` methods/api are implemented in jsdom so I still don't know what are the limits without those API.
I am having difficulties in sorting things to make them work with just native functionality.
But I am not defeat yet !
I will have a look to some previous suggestion by @TimothyGu on the same matter (localName vs. nodeName properties and namespaceURI too).

**Comment by domenic** [comment]
If implementing those methods would help, we can do that! I didn't even know they existed, haha.

**Comment by dperini** [comment]
I am not sure yet ... maybe I can do without them, in a couple of days I will have more bits in that.
Anyway when you run out of the "todo list" here are the starting point in pseudo code:

    https://www.w3.org/TR/DOM-Level-3-Core/namespaces-algorithms.html

also be known that some of these maybe already obsoleted in browsers:

https://www.w3.org/TR/DOM-Level-3-Core/core.html#Node3-lookupNamespacePrefix

**Comment by domenic** [comment]
Right, we would implement them per https://dom.spec.whatwg.org/#dom-node-lookupprefix.

**Comment by dperini** [comment]
@domenic here you can find more related code samples that might help:

https://developer.mozilla.org/en-US/docs/Archive/Add-ons/Code_snippets/LookupPrefix

**Comment by Zirro** [comment]
This continues in https://github.com/jsdom/jsdom/pull/2229.

**Pull request closed without merging**